### PR TITLE
New links

### LINKED ISSUE #34: Refactor links
Need to go over our link components, and make sure they work as expected, have docs, compositions, etc.

Specifics:
- Make sure we have a working Navigation Link, on all levels.
- Make sure we have a working react implementation.
- Make sure we can have a virtual base name, so we can prepend "https://bit.dev" to our scope links.
- make a url builder hook `useHref()`, so we can dynamically build urls for things like preview iframes, etc.

Verify:
- [ ] component env is default "react"

Then: 
- [ ] use new components in bit-bin
- [ ] deprecate and remove all traces of the older links.

### PR DESCRIPTION
**Pull request by KutnerUri**
closed #34

- Go over **all** of the link components
- Make sure they work in all conditions
- Separate `NavLink` (simplify usage)
- Add docs, compositions, verify everything works
- Support virtual base name for React Router adapter. (so we can add `http://bit.dev` as to the scopes in bit.dev)

TODO
- [x] fix warning about forwardRef
- [ ] decide on env (currently `community-react`)
- [x] verify all docs are correct
- [x] verify tests, bit build

### CODE REVIEW

File: base-react/navigation/router-context/router-context-provider.tsx [hunk 1]
```diff
@@ -1,23 +1,31 @@
...
-  );
-}
+// export function RouterContextProvider({
+//   children,
+//   ...rest
+// }: RouterContextProviderProps) {
+//   return (
+//     <RouterContext.Provider value={rest}>{children}</RouterContext.Provider>
+//   );
+// }
```
Comment (odedre) [inline base-react/navigation/router-context/router-context-provider.tsx:31]: cleanup

File: community/apps/bit-dev/app.tsx [hunk 2]
```diff
@@ -26,46 +28,45 @@ export function BitDevApp() {
   const [highlighting, setHighlighting] = useState(true);
 
   return (
+    // {/* TODO @Uri - remove the legacy RoutingProvider */}
     <RoutingProvider value={legacyRouting}>
       <SidebarProvider>
-        <RouterContextProvider Link={Link}>
+        <ReactRouterRoutingAdapter useBrowserRouter>
```
Comment (odedre) [inline community/apps/bit-dev/app.tsx:34]: consider renaming `ReactRouterRoutingAdapter`. maybe to `ReactRouterProvider`.
I think its a clearer name

File: community/ui/header/nav/nav-links.tsx [hunk 3]
```diff
@@ -1,10 +1,10 @@
 import React, { ReactNode } from 'react';
 // import { TooltipDrawer } from '@teambit/evangelist.surfaces.tooltip';
-import classNames from 'classnames';
-import { ExternalLink } from '@teambit/design.ui.external-link';
+// import classNames from 'classnames';
+// import { ExternalLink } from '@teambit/design.ui.external-link';
```
Comment (odedre) [inline community/ui/header/nav/nav-links.tsx:4]: cleanup

File: ui-foundation/ui/navigation/react-router/link-anchor/link-anchor.base.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
+/* eslint-disable prefer-const */
+/* eslint-disable react/prop-types */
+
+// This is a clone of React Router's AnchorLink, which is otherwise not public.
+// It was copied 'as is' from here:
+// https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/modules/Link.js
```
Comment (odedre) [inline ui-foundation/ui/navigation/react-router/link-anchor/link-anchor.base.js:6]: jesus

Comment (odedre) [reply ui-foundation/ui/navigation/react-router/link-anchor/link-anchor.base.js:6]: I think we need to add it to the list of forked stuff. talk to guy about it

Comment (odedre) [reply ui-foundation/ui/navigation/react-router/link-anchor/link-anchor.base.js:6]: also, maybe we should open an issue in react router

File: ui-foundation/ui/navigation/react-router/link/link.tsx [hunk 5]
```diff
@@ -1,32 +1,57 @@
...
+  const component = baseUrl ? LinkAnchor : undefined;
+
+  // @ts-ignore (https://github.com/teambit/bit/issues/4401)
+  return <BaseLink to={to} {...rest} component={component} />;
+}
+
+/**
+ * Adapter between React router's NavLink and our isomorphic Navlink components. Learn more [Here](https://bit.dev/teambit/base-react/navigation/router-context)
+ */
+export function NavLink({
```
Comment (odedre) [inline ui-foundation/ui/navigation/react-router/link/link.tsx:33]: consider separating to different file

File: workspace.jsonc [hunk 6]
```diff
@@ -126,16 +129,20 @@
...
       },
       "peerDependencies": {
         "@testing-library/react": "12.0.0",
+        "@types/history": "^4.0.0",
+        "@types/react-router-dom": "^5.0.0",
+        "history": "^5.0.0",
         "react": "16.13.1",
-        "react-dom": "16.13.1"
+        "react-dom": "16.13.1",
+        "react-router-dom": "^5.0.0"
```
Comment (odedre) [inline workspace.jsonc:145]: why did you downgrade the version?

**Review by odedre — APPROVED** [review]

File: workspace.jsonc (hunk 6 above)
Comment (KutnerUri) [reply workspace.jsonc:145]: I did not downgrade, just set it to be any version > 5.0.0, so we'd support a large range

File: ui-foundation/ui/navigation/react-router/link/link.tsx (hunk 5 above)
Comment (KutnerUri) [reply ui-foundation/ui/navigation/react-router/link/link.tsx]: done

File: ui-foundation/ui/navigation/react-router/link-anchor/link-anchor.base.js (hunk 4 above)
Comment (KutnerUri) [reply ui-foundation/ui/navigation/react-router/link-anchor/link-anchor.base.js:6]: 👌

File: community/apps/bit-dev/app.tsx (hunk 2 above)
Comment (KutnerUri) [reply community/apps/bit-dev/app.tsx]: OK

File: base-react/navigation/router-context/router-context-provider.tsx (hunk 1 above)
Comment (KutnerUri) [reply base-react/navigation/router-context/router-context-provider.tsx]: ok

File: base-react/navigation/router-context/link.type.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+   * key-value state for stateful routing systems. (Has to be supported by the speicifc link implementation)
+   */
+  state?: Record<string, any>;
+
+  /**
+   * force the use of a native `a` element and ignore the contextual implemented Link.
+   */
+  native?: boolean;
+
+  external?: boolean;
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply base-react/navigation/router-context/link.type.ts:15]: done

**Pull request merged**